### PR TITLE
Add avoid non null assertion rule

### PR DESCRIPTION
**Pull request by incendial**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix
[X] New rule
[ ] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

### CODE REVIEW

**Comment by roman-petrov** [comment]
Just got an email notification about this PR and I think it is great rule! I already seen a feature request asking to implement exactly the same in dart linter repository 

[Lint rule for ! postfix bang/exclamation and late null safety language features](https://github.com/dart-lang/linter/issues/2562)

**Comment by incendial** [comment]
Yeah, I've actually seen this issue and decided to add this feature. Unfortunately there is no rule requests in our repo yet, hope this will change in the future.

File: lib/src/obsoleted/rules/avoid_non_null_assertion_rule.dart [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    source.unit?.visitChildren(visitor);
+
+    return visitor.expressions
+        .map((expression) => createIssue(
+              rule: this,
+              location: nodeLocation(
+                node: expression,
+                source: source,
+                withCommentOrMetadata: true,
```
Comment (dkrutskikh) [inline lib/src/obsoleted/rules/avoid_non_null_assertion_rule.dart:38]: @incendial please remove `withCommentOrMetadata: true,`

**Review by dkrutskikh — CHANGES_REQUESTED** [review]

File: lib/src/obsoleted/rules/avoid_non_null_assertion_rule.dart (hunk 1 above)
Comment (incendial) [reply lib/src/obsoleted/rules/avoid_non_null_assertion_rule.dart]: Done.

**Pull request merged**